src/testRunner/unittests/tsbuild.ts

```diff
@@ -455,6 +456,122 @@ export const b = new A();`);
+                fs.writeFileSync(tsconfigBaseFile, JSON.stringify(tsconfigBase));
+
+                const lengthFile = "/src/converter/length.ts";
+                const lengthContent = fs.readFileSync(lengthFile, "utf8");
+                fs.writeFileSync(lengthFile, `import { leftPad } from "utils";
+import { foo } from "../utils/someFile";
+
+${lengthContent}`);
+
+                fs.writeFileSync("/src/utils/types.d.ts", `declare module "utils" {
+    export function leftPad(s: string, n: number): string;
+}`);
+
+                const someFile = "/src/utils/someFile.ts";
+                const someFileContent = fs.readFileSync(someFile, "utf8");
+                fs.writeFileSync(someFile, `export ${someFileContent}`);
+
+                const unUsedFile = "/src/utils/unusedFile.ts";
+                const unUsedFileContent = fs.readFileSync(unUsedFile, "utf8");
+                fs.writeFileSync(unUsedFile, `export ${unUsedFileContent}`);
+            }
+
+            function makeOutFile(fs: vfs.FileSystem) {
+                const tsconfigFile = "/src/utils/tsconfig.json";
+                const tsconfig = JSON.parse(fs.readFileSync(tsconfigFile, "utf8"));
+                tsconfig.compilerOptions = { outFile: "result.js" };
+                fs.writeFileSync(tsconfigFile, JSON.stringify(tsconfig));
+            }
+
+            describe("Non modules", () => {
+                it("without --out", () => {
+                    verifyDtsIncluded(noop, allExpectedOutputs, [
+                        ...expectedFileTracesUtils,
+                        ...getLibs(),
+                        "/src/utils/someFile.d.ts", // output d.ts
+                        "/src/utils/types.d.ts", // source d.ts
+                        "/src/utils/unusedFile.d.ts", // output d.ts
+                        "/src/converter/length.ts"
+                    ]);
+                });
+
+                it("with --out", () => {
+                    verifyDtsIncluded(fs => makeOutFile(fs), allExpectedOutputsWithOutFile, [
+                        ...expectedFileTracesUtils,
+                        ...getLibs(),
+                        "/src/utils/result.d.ts", // out file
+                        "/src/utils/types.d.ts", // d.ts should be included by default
+                        "/src/converter/length.ts"
+                    ]);
+                });
+            });
+
+            describe("modules", () => {
+                it("without --out", () => {
+                    verifyDtsIncluded(fs => makeModules(fs), allExpectedOutputs, [
+                        ...expectedFileTracesUtils,
+                        ...getLibs(),
+                        "/src/utils/types.d.ts", // d.ts should be included by default
+                        "/src/utils/someFile.d.ts", // As part of module resolution
+                        "/src/converter/length.ts"
+                    ]);
+                });
+
+                it.skip("with --out", () => {

```

Comment:
This is issue unrelated to this PR and exists even now given that the current assumption that 1 source file = 1 generated file for module scenario is not true with --out and amd and we end up including d.ts multiple times. This should be fixed in different PR and hence the test is skipped.

Reply:
Fixed with 5d3e2ee